### ISSUE TITLE
Admin impersonation of other users

### ISSUE THREAD

**Issue by vygandas** [cortip/isomera#174]
It should be possible to impersonate users as master admin (owner of the project). It's quite common that admins need to help users to do something and that's almost in all projects I worked on. Let's save more time for users by adding this feature.

**Comment by hirenf14** [comment]
Currently,
In `api`,
I can see Organizations, and Users have specific role (Owner / Member) in the Organization.

1. Should Organization owner be able to impersonate users from their organization?
2. For master admin, we need to define identification, (conventional way: is_admin column in users table).

is it correct?

Thanks.

**Comment by vygandas** [comment]
Hello @hirenf14, 

Regarding your questions:

1. No. I think that could bring too many privacy issues.
2. Yes, that sounds good. Here, the super admin could help users solve issues.

Cheers

**Comment by hirenf14** [comment]
Hi @vygandas ,

There isn't any users listing, should it be available for super admin when they login? Or we can just keep it in API for now?

**Comment by vygandas** [comment]
@hirenf14 at this moment, we do not have any admin area to list users. So, there is no starting point for this feature 🤔 One idea - the admin could click a button where they would need to enter the user ID (or even better - generated temp PIN) + email and then could impersonate. I'd imagine a client calling on the phone and telling these details so admins couldn't randomly impersonate clients. Or maybe it could be done like Godaddy has done it for support tickets? That would be convenient for most projects, I guess?

**Comment by hirenf14** [comment]
Yes, that can be a good way to implement it, only issue would be temporary PIN logic might be deeper then it may seem. Considering we might need to allow users' to generate these PIN, and invalidating them.
Haven't use Godaddy much, so what's their idea for this?
From how I look into it,
1. User allows to impersonate someone.
2. For Admin, UI need to indicate that its personation.
3. A way to stop/quit impersonation (2nd API to revert the token)

On another note, I checked, there are common functions to validate / regenerate tokens but with 2FA logic, they were ignored (including old jwt strategy is not useful anymore as payload is wrong in it self).

**Comment by vygandas** [comment]
> Haven't use Godaddy much, so what's their idea for this?

Here is how it looks:
![image](https://github.com/cortip/isomera/assets/723200/55c37241-dbb3-4d1c-aaf0-ca1afe79a260)

The support agent would ask for the client's ID and PIN. The pin is rotating. I think it changes every login or on some timer.

> On another note, I checked, there are common functions to validate / regenerate tokens but with 2FA logic, they were ignored (including old jwt strategy is not useful anymore as payload is wrong in it self).

That sounds like a bug 🤔
Probably this should be addressed first haha.

---

Workflow you described makes sense. Tho the question is - how to differentiate Admins UI with least changes to overall UI.
Since this is a starter project, we should not go too far and create too many different pages. Maybe if it's admin, this feature could appear somewhere where user would look for settings or logout? You know, when click on top right corner avatar and menu drops.

**Comment by hirenf14** [comment]
Easiest way to address UI identification is a Banner at the top, with info on impersonation, For ex: "Viewing as John Doe (close)". On close, it can quit the impersonation.
Agree with Admin suggestion as well.
